diff --git a/src/ap/neighbor_db.c b/src/ap/neighbor_db.c
index 413df79..de31381 100644
--- a/src/ap/neighbor_db.c
+++ b/src/ap/neighbor_db.c
@@ -411,6 +411,10 @@ int add_stapro_neighbor_rep(struct hostapd_data *hapd, u8 *nei_rep,
 
 	if (!hapd || !sm)
 		return nei_len;
+	if (!sm->sta_ml_ie) {
+		wpa_printf(MSG_DEBUG, "sta_ml_ie is NULL!!\n");
+		return nei_len;
+	}
 	memcpy(&ml, sm->sta_ml_ie, sizeof(struct wpa_ml_ie_parse));
 	ml_link_num = (int)(nei_len / 29);   //gives total neighbours present in nei_rep
 	if (ml_link_num <= 1)
