diff --git a/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
index ab681276a..a805b1262 100644
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -3930,6 +3930,10 @@ static void handle_auth(struct hostapd_data *hapd,
 #endif /* CONFIG_SAE */
 
 	sta = ap_get_sta(hapd, mgmt->sa);
+	if (sta == NULL && auth_alg == WLAN_AUTH_SAE && auth_transaction == 2) {
+		wpa_printf(MSG_ERROR, "Comfirm without commit is not allowed\n");
+		return;
+	}
 	if (sta) {
 		sta->flags &= ~WLAN_STA_PENDING_FILS_ERP;
 		sta->ft_over_ds = 0;
