diff --git a/src/ml/ml_common.c b/src/ml/ml_common.c
index be8fdd8..e916a28 100644
--- a/src/ml/ml_common.c
+++ b/src/ml/ml_common.c
@@ -1245,17 +1245,32 @@ u8* ml_add_rekey_kde(struct wpa_state_machine *sm, u8 *pos)
 {
 	struct wpa_ml_group *ml_group = NULL;
 	struct wpa_ml_link *link;
-	u8 i;
+	u8 i, j;
 
 	if (!sm->dot11MultiLinkActivated)
 		return pos;
 
 	wpa_printf(MSG_INFO, "ML: Add Mac/GTK into EAPOL-Key rekey");
+
 	ml_group = STATE_MACHINE_ML_GROUP;
 	pos = ml_set_mac_kde(pos, ml_group->ml_addr);
 
 	for (i = 0; i < ml_group->ml_link_num; i++) {
+		u8 found = false;
+
 		link = &ml_group->links[i];
+		if (link->link_id == sm->sta_ml_ie->link_id) {
+			found = true;
+		} else {
+			for (j = 0; j < sm->sta_ml_ie->prof_num; j++) {
+				if (sm->sta_ml_ie->profiles[j].link_id == link->link_id)
+					found = true;
+			}
+		}
+
+		if (!found)
+			continue;
+
 		pos = ml_set_gtk_kde(sm, pos, link);
 		pos = ml_set_ieee80211w_kde(sm, pos, link);
 	}
