diff --git a/src/ap/ap_config.h b/src/ap/ap_config.h
index 7545548..fbd0a04 100644
--- a/src/ap/ap_config.h
+++ b/src/ap/ap_config.h
@@ -19,6 +19,7 @@
 #include "wps/wps.h"
 #include "fst/fst.h"
 #include "vlan.h"
+#include "ml/ml_common.h"
 
 enum macaddr_acl {
 	ACCEPT_UNLESS_DENIED = 0,
@@ -961,6 +962,7 @@ struct hostapd_bss_config {
 #ifdef CONFIG_IEEE80211AX
 	u8 dot11vmbssid;
 #endif /* CONFIG_IEEE80211AX */
+	struct wpa_ml_ie_parse ml_ie_btm;
 };
 
 /**
diff --git a/src/ap/ctrl_iface_ap.c b/src/ap/ctrl_iface_ap.c
index 34ff7bf..1e9a0f8 100644
--- a/src/ap/ctrl_iface_ap.c
+++ b/src/ap/ctrl_iface_ap.c
@@ -1155,6 +1155,7 @@ int hostapd_ctrl_iface_bss_tm_req(struct hostapd_data *hapd,
 	int nei_len;
 	u8 mbo[10];
 	size_t mbo_len = 0;
+	int i = 0;
 
 	if (hwaddr_aton(cmd, addr)) {
 		wpa_printf(MSG_DEBUG, "Invalid STA MAC address");
@@ -1162,6 +1163,20 @@ int hostapd_ctrl_iface_bss_tm_req(struct hostapd_data *hapd,
 	}
 
 	sta = ap_get_sta(hapd, addr);
+	if (sta == NULL) {
+		for (i = 0; i < 3; i++) {
+			struct hostapd_iface *iface = hapd->iface;
+			if (iface) {
+				struct hostapd_iface *h = iface->interfaces->iface[i];
+				struct hostapd_data *h_hapd = h->bss[0];
+				if(!h_hapd)
+					continue;
+				sta = ap_get_sta(h_hapd, addr);
+				if (sta)
+					break;
+			}
+		}
+	}
 	if (sta == NULL) {
 		wpa_printf(MSG_DEBUG, "Station " MACSTR
 			   " not found for BSS TM Request message",
@@ -1215,6 +1230,8 @@ int hostapd_ctrl_iface_bss_tm_req(struct hostapd_data *hapd,
 
 	nei_len = ieee802_11_parse_candidate_list(cmd, nei_rep,
 						  sizeof(nei_rep));
+	nei_len = add_stapro_neighbor_rep(hapd, nei_rep, nei_len);
+	wpa_hexdump(MSG_DEBUG, "nei_rep with per sta", nei_rep, nei_len);
 	if (nei_len < 0)
 		return -1;
 
diff --git a/src/ap/neighbor_db.c b/src/ap/neighbor_db.c
index c73dd6b..4ac14dd 100644
--- a/src/ap/neighbor_db.c
+++ b/src/ap/neighbor_db.c
@@ -400,6 +400,128 @@ int hostapd_neighbor_check_other_mac_presence(char *nei_buf, size_t buflen, u8 *
 	return own_mac;
 }
 
+int add_stapro_neighbor_rep(struct hostapd_data *hapd, u8 *nei_rep,
+		size_t nei_len)
+{
+	int len = 0, shift_len = 0, temp_len = 0, sta_pro_len = 0, i = 0, j = 0, ap_nei = 0;
+	u8 *pos = nei_rep;
+	u8 *start_pro, *next_nei_rep;
+	size_t ml_link_num = hapd->ml_group->ml_link_num;
+	struct wpa_ml_ie_parse ml = hapd->conf->ml_ie_btm;
+
+	if (!hapd)
+		return nei_len;
+	ml_link_num = (int)(nei_len / 29);   //gives total neighbours present in nei_rep
+	if (ml_link_num <= 1)
+		return nei_len;
+	temp_len = 2 + 6 + 4 + 1 + 1 + 1 + 3 + 11;   //length of neighbour_report till common info(without linkid)
+	shift_len = nei_len / ml_link_num;          // linkid is included or not, if shift_len > temp_len
+
+	while (len < nei_len && j < ml_link_num) {    //AP's own links as neighbour, ap_nei
+		if (nei_rep[len] == WLAN_EID_NEIGHBOR_REPORT) {
+			struct hostapd_iface *iface = hapd->iface;
+			if (iface) {
+				struct hostapd_iface *h = iface->interfaces->iface[i];
+				struct hostapd_data *h_hapd = h->bss[0];
+				if (!h_hapd) {
+					j++;
+					continue;
+				}
+				if ((os_memcmp(&nei_rep[len + 2], ml.profiles[i].addr, ETH_ALEN) == 0) ||
+					(os_memcmp(&nei_rep[len + 2], h_hapd->own_addr, ETH_ALEN) == 0)) {
+					if (shift_len == temp_len)  //if link_id not present, no need to add per_sta profile
+						return nei_len;
+					ap_nei++;
+					len += temp_len + 1;
+					j++;
+					i = 0;
+				} else {
+					i++;
+					if (i > 2) {
+						i = 0;
+						j++;
+						len += temp_len;
+					}
+				}
+			} else
+				break;
+		} else {
+			wpa_printf(MSG_DEBUG, "Invalid report\n");
+			break;
+		}
+	}
+	if (ap_nei < 2)      //if single then no need to add per-sta
+		return nei_len;
+	len = 0;
+	i = 0;
+	j = 0;
+	while (len < nei_len && j < ml_link_num) {
+		if (nei_rep[len] == WLAN_EID_NEIGHBOR_REPORT) {
+			struct hostapd_iface *iface = hapd->iface;
+			if (iface) {
+				struct hostapd_iface *h = iface->interfaces->iface[i];
+				struct hostapd_data *h_hapd = h->bss[0];
+				if (!h_hapd) {
+					j++;
+					continue;
+				}
+				if ((os_memcmp(&nei_rep[len + 2], ml.profiles[i].addr, ETH_ALEN) == 0) ||
+					(os_memcmp(&nei_rep[len + 2], h_hapd->own_addr, ETH_ALEN) == 0)){
+					u16 sta_ctrl = 0;
+					int other_id1 = 0, other_id2 = 0, ap_nei_tmp = ap_nei;
+					pos = pos + temp_len + 1;
+					start_pro = pos;
+					if (j < ml_link_num - ap_nei_tmp) {        //nei_rep of all the neighbours that are not Ap's
+						next_nei_rep = os_malloc(temp_len * (ml_link_num - (j + ap_nei_tmp)));
+						os_memcpy(next_nei_rep, (start_pro + (ap_nei_tmp - 1) * (temp_len + 1)),
+							temp_len * (ml_link_num - (j + ap_nei_tmp)));
+					}
+					other_id1 = *(pos + temp_len);    //link_id of AP's second nei
+					if (ap_nei > 2)             //link_id of AP's third nei
+						other_id2 = *(pos + (2 * temp_len) + 1);
+					while (ap_nei > 1) {
+						*pos++ = ML_SUB_ID_PER_STA_PROFILE;
+						*pos++ = 2;
+						if (other_id1 >= 0) {
+							sta_ctrl |= other_id1 & 0x0f;
+							other_id1 = -1;
+						} else
+							sta_ctrl |= other_id2 & 0x0f;
+						sta_ctrl |= 0 << 4;  //ml.profiles[i].complete_profile << 4;
+						sta_ctrl |= 0 << 5;  //ml.profiles[i].mac_addr_present << 5;
+						sta_ctrl |= 0 << 6;  //ml.profiles[i].bcn_intvl_present << 6;
+						sta_ctrl |= 0 << 7;  //ml.profiles[i].tsf_offset_present << 7;
+						sta_ctrl |= 0 << 8;  //ml.profiles[i].dtim_present << 8;
+						sta_ctrl |= 0 << 9;  //ml.profiles[i].nstr_link_pair_present << 9;
+						sta_ctrl |= 0 << 10; //ml.profiles[i].nstr_present << 10;
+						sta_ctrl |= 0 << 11; //ml.profiles[i].bss_para_change_count_present << 11;
+						WPA_PUT_LE16(pos, sta_ctrl);
+						pos += 2;
+						ap_nei--;
+					}
+					if (j < ml_link_num - ap_nei_tmp)
+						memcpy (pos, next_nei_rep, temp_len * (ml_link_num - (j + ap_nei_tmp)));
+					return ((nei_len + pos - start_pro) - ((ap_nei_tmp - 1) * (temp_len + 1)));
+				} else {
+					i++;
+					if (i > 2) {
+						wpa_printf(MSG_DEBUG, "This mac is not in list\n");
+						len += temp_len;
+						pos += temp_len;
+						j++;
+						i = 0;
+					}
+				}
+			} else
+				break;
+		} else {
+			wpa_printf(MSG_DEBUG, "Invalid report\n");
+			break;
+		}
+	}
+	return nei_len;
+}
+
 void hostapd_neighbor_set_own_report_pref(struct hostapd_data *hapd, char *nei_buf,
 		size_t buflen, const int pref)
 {
diff --git a/src/ap/neighbor_db.h b/src/ap/neighbor_db.h
index 1f8bd6a..8f69c37 100644
--- a/src/ap/neighbor_db.h
+++ b/src/ap/neighbor_db.h
@@ -27,6 +27,7 @@ void hostapd_free_neighbor_db(struct hostapd_data *hapd);
 int hostapd_neighbor_count(struct hostapd_data *hapd);
 int hostapd_neighbor_insert_buffer(struct hostapd_data *hapd, char *buf,
         size_t buflen);
+int add_stapro_neighbor_rep(struct hostapd_data *hapd, u8 *nei_rep, size_t nei_len);
 void hostapd_neighbor_set_own_report_pref(struct hostapd_data *hapd, char *nei_buf,
 		                size_t buflen, const int pref);
 #ifdef CONFIG_MBO
diff --git a/src/ap/wpa_auth_ft.c b/src/ap/wpa_auth_ft.c
index a8dea61..f9cbbca 100644
--- a/src/ap/wpa_auth_ft.c
+++ b/src/ap/wpa_auth_ft.c
@@ -3694,7 +3694,7 @@ static int wpa_ft_process_auth_req(struct wpa_state_machine *sm,
 		const u8 *ml_ie = parse.mlie;
 		size_t ml_ie_len = parse.mlie_len;
 
-		if (ml_parse_ie(ml_ie, ml_ie_len, &ml, NULL) != 0) {
+		if (ml_parse_ie(ml_ie, ml_ie_len, &ml, NULL, NULL) != 0) {
 			sm->sta_ml_ie = NULL;
 			sm->dot11MultiLinkActivated = 0;
 			wpa_printf(MSG_DEBUG, "FT: Invalid MLIE");
diff --git a/src/common/wpa_common.c b/src/common/wpa_common.c
index 5c0f6f0..a86932b 100644
--- a/src/common/wpa_common.c
+++ b/src/common/wpa_common.c
@@ -950,7 +950,7 @@ int wpa_ft_mic(int key_mgmt, const u8 *kck, size_t kck_len, const u8 *sta_addr,
 
 #ifdef CONFIG_MTK_IEEE80211BE
 	if (mlie) {
-		if (ml_parse_ie(mlie, mlie_len, &ml_info, NULL) == 0) {
+		if (ml_parse_ie(mlie, mlie_len, &ml_info, NULL, NULL) == 0) {
 			valid_links |= BIT(ml_info.link_id);
 			link_addr[ml_info.link_id] = main_link_addr;
 			link_rsne[ml_info.link_id] = rsnie;
@@ -1518,7 +1518,7 @@ int wpa_ft_parse_ies(const u8 *ies, size_t ies_len,
 				wpa_hexdump(MSG_DEBUG, "FT: ML IE", pos + 1, (unsigned int) len - 1);
 
 				if (ml_parse_ie(parse->mlie, parse->mlie_len,
-					        &ml, NULL) == 0) {
+					        &ml, NULL, NULL) == 0) {
 					link_num = ml.prof_num + 1;
 					wpa_printf(MSG_INFO, "FT: ML total link num %d", link_num);
 				}
diff --git a/src/drivers/mediatek_driver_cmd_nl80211.c b/src/drivers/mediatek_driver_cmd_nl80211.c
index 9ce28c1..2ad5d2e 100644
--- a/src/drivers/mediatek_driver_cmd_nl80211.c
+++ b/src/drivers/mediatek_driver_cmd_nl80211.c
@@ -426,7 +426,7 @@ void mtk_nl80211_rxt2lm_stop_disassoc_timer_event(struct wpa_driver_nl80211_data
 				prev = sta;
 				if (sta->flags & WLAN_STA_ASSOC) {
 					wpa_printf(MSG_DEBUG,"cancelling timer for STA" MACSTR, MAC2STR(sta->addr));
-					eloop_cancel_timeout(ap_handle_timer, h_hapd, sta);
+					eloop_cancel_timeout(ap_handle_timer, ELOOP_ALL_CTX, sta);
 				}
 				sta = sta->next;
 			}
diff --git a/src/ml/ml_common.c b/src/ml/ml_common.c
index d382129..a3957b6 100644
--- a/src/ml/ml_common.c
+++ b/src/ml/ml_common.c
@@ -190,7 +190,7 @@ u8* ml_set_ml_rsno_link_kde(u8 *pos, u8 link_id,
 	return pos;
 }
 
-int ml_parse_ie(const u8 *ie, size_t len, struct wpa_ml_ie_parse *ml, u8 *bssid)
+int ml_parse_ie(const u8 *ie, size_t len, struct wpa_ml_ie_parse *ml, u8 *bssid, struct hostapd_data *hapd)
 {
 	const u8 *pos, *end, *ci_end, *info_end;;
 	u16 ml_ctrl;
@@ -306,6 +306,14 @@ int ml_parse_ie(const u8 *ie, size_t len, struct wpa_ml_ie_parse *ml, u8 *bssid)
 		wpa_printf(MSG_INFO, "ML:common Info MLD id = %u", ml->mld_id);
 		pos += 1;
 	}
+	if (ml_ctrl & ML_CTRL_EXT_MLD_CAP_PRESENT) {
+		if (ci_end - pos < 2) {
+			wpa_printf(MSG_ERROR,
+					"No room for Extended MLD Cap and Operations of Commmon Info");
+			return -1;
+		}
+		pos += 2;
+	}
 
 	if (pos - (ie + 2) != ml->common_info_len) {
 		ml->valid = false;
@@ -377,6 +385,23 @@ int ml_parse_ie(const u8 *ie, size_t len, struct wpa_ml_ie_parse *ml, u8 *bssid)
 			profile->mac_addr_present = 1;
 			wpa_printf(MSG_INFO, "ML: LinkID=%u, LinkAddr="MACSTR"",
 				profile->link_id, MAC2STR(profile->addr));
+			if (hapd) {
+				struct hostapd_iface *iface = hapd->iface;
+				if (iface) {
+					struct hostapd_iface *h = iface->interfaces->iface[ml->prof_num];
+					struct hostapd_data *h_hapd = h->bss[0];
+					struct sta_info *sta = NULL;
+					sta = ap_get_sta(h_hapd, profile->addr);
+					if (sta)
+						continue;
+					sta = ap_sta_add(h_hapd, profile->addr);
+					if (sta) {
+						sta->flags |= WLAN_STA_AUTH | WLAN_STA_ASSOC;
+						os_memset(hapd->profiles[ml->prof_num].addr, 0, ETH_ALEN);
+						os_memcpy(hapd->profiles[ml->prof_num].addr, profile->addr, ETH_ALEN);
+					}
+				}
+			}
 			pos += ETH_ALEN;
 		}
 		if (sta_ctrl & ML_STA_CTRL_BCN_INTV_PRESENT) {
@@ -473,7 +498,7 @@ int ml_parse_ie(const u8 *ie, size_t len, struct wpa_ml_ie_parse *ml, u8 *bssid)
 		 */
 		while (frag_flag && (pos + 1) < end) {
 			u8 frag_len = 0;
-
+			pos += 2;
 			if (*pos != 254) {
 				wpa_printf(MSG_INFO,
 					"ML: invalid sta profile FragID(254) %u", *pos);
@@ -534,7 +559,7 @@ int ml_sae_process_auth(struct sae_data *sae, u16 auth_transaction,
 			sae->peer_ml_ie = NULL;
 			sae->dot11MultiLinkActivated = 0;
 		} else {
-			if (ml_parse_ie(ml_ie, ml_ie_len, &ml, NULL) != 0) {
+			if (ml_parse_ie(ml_ie, ml_ie_len, &ml, NULL, NULL) != 0) {
 				sae->peer_ml_ie = NULL;
 				sae->dot11MultiLinkActivated = 0;
 				if (frag_ml_ie)
@@ -566,7 +591,7 @@ int ml_sae_process_auth(struct sae_data *sae, u16 auth_transaction,
 		}
 
 		if (ml_ie) {
-			if (ml_parse_ie(ml_ie, ml_ie_len, &ml, NULL) != 0) {
+			if (ml_parse_ie(ml_ie, ml_ie_len, &ml, NULL, NULL) != 0) {
 				wpa_printf(MSG_ERROR, "ML: SAE confirm failed to parse ml ie");
 				if (frag_ml_ie)
 					wpabuf_free(frag_ml_ie);
@@ -892,6 +917,10 @@ int ml_new_assoc_sta(struct wpa_state_machine *sm, const u8 *ie, size_t len)
 {
 	if (!sm)
 		return -1;
+	struct hostapd_data *hapd = (struct hostapd_data *)sm->wpa_auth->cb_ctx;
+	if (!hapd)
+		return -1;
+
 	wpa_printf(MSG_INFO, "ML: new STA:");
 	os_free(sm->sta_ml_ie);
 	if (ie == NULL || len == 0 || STATE_MACHINE_ML_GROUP == NULL) {
@@ -901,7 +930,7 @@ int ml_new_assoc_sta(struct wpa_state_machine *sm, const u8 *ie, size_t len)
 		struct wpa_ml_ie_parse ml;
 		struct wpa_ml_group *ml_group = STATE_MACHINE_ML_GROUP;
 
-		if (ml_parse_ie(ie, len, &ml, NULL) != 0) {
+		if (ml_parse_ie(ie, len, &ml, NULL, hapd) != 0) {
 			sm->sta_ml_ie = NULL;
 			sm->dot11MultiLinkActivated = 0;
 			return -1;
diff --git a/src/ml/ml_common.h b/src/ml/ml_common.h
index 04b434d..e905945 100644
--- a/src/ml/ml_common.h
+++ b/src/ml/ml_common.h
@@ -81,6 +81,7 @@ struct sae_data;
 #define ML_CTRL_EML_CAPA_PRESENT			BIT(7)
 #define ML_CTRL_MLD_CAPA_PRESENT			BIT(8)
 #define ML_CTRL_MLD_ID_PRESENT				BIT(9)
+#define ML_CTRL_EXT_MLD_CAP_PRESENT                     BIT(10)
 
 #define ML_SUB_ID_PER_STA_PROFILE			0
 #define ML_STA_CTRL_LINK_ID_MASK			BITS(0, 3)
@@ -194,7 +195,7 @@ const u8 * ml_auth_aa(struct wpa_state_machine *sm, const u8 *bssid);
 u8* ml_set_mac_kde(u8 *buf, const unsigned char *addr);
 u8* ml_set_ml_link_kde(u8 *pos, u8 id, const unsigned char *addr,
 	const u8 *rsne, size_t rsne_len, const u8 *rsnxe, size_t rsnxe_len);
-int ml_parse_ie(const u8 *pos, size_t len, struct wpa_ml_ie_parse *ml, u8 *bssid);
+int ml_parse_ie(const u8 *pos, size_t len, struct wpa_ml_ie_parse *ml, u8 *bssid, struct hostapd_data *hapd);
 
 #ifdef CONFIG_SAE
 int ml_sae_process_auth(struct sae_data *sae, u16 auth_transaction,
diff --git a/src/ml/ml_supplicant.c b/src/ml/ml_supplicant.c
index 207ca2e..db837ca 100644
--- a/src/ml/ml_supplicant.c
+++ b/src/ml/ml_supplicant.c
@@ -250,7 +250,7 @@ int ml_set_sae_auth_commit_req_ml_ie (struct sae_data *sae, const u8 *ies, size_
 		wpa_printf(MSG_DEBUG, "ML: clearing STA ML IE");
 		sae->dot11MultiLinkActivated = 0;
 	} else {
-		if (ml_parse_ie(ml_ie, ml_ie_len, &ml, NULL) != 0) {
+		if (ml_parse_ie(ml_ie, ml_ie_len, &ml, NULL, NULL) != 0) {
 			sae->dot11MultiLinkActivated = 0;
 			if (frag_ml_ie)
 				wpabuf_free(frag_ml_ie);
@@ -307,7 +307,7 @@ int ml_set_assoc_req_ml_ie(struct wpa_sm *sm, const u8 *ies, size_t ies_len)
 		wpa_printf(MSG_DEBUG, "ML:[%s] clearing STA ML IE", __func__);
 		goto err;
 	} else {
-		if (ml_parse_ie(ml_ie, ml_ie_len, &ml, NULL) != 0 ||
+		if (ml_parse_ie(ml_ie, ml_ie_len, &ml, NULL, NULL) != 0 ||
 		    ml.prof_num > ML_MAX_LINK_NUM) {
 			wpa_printf(MSG_DEBUG, "ML:[%s] ml_parse_ie fails", __func__);
 			goto err;
@@ -376,7 +376,7 @@ int ml_set_assoc_resp_ml_ie(struct wpa_sm *sm, const u8 *ies, size_t ies_len, u8
 	} else {
 		WPA_ASSERT(sm->dot11MultiLinkActivated);
 
-		if (ml_parse_ie(ml_ie, ml_ie_len, &ml, bssid) != 0  ||
+		if (ml_parse_ie(ml_ie, ml_ie_len, &ml, bssid, NULL) != 0  ||
 		    ml.prof_num > ML_MAX_LINK_NUM) {
 			wpa_printf(MSG_DEBUG, "ML:[%s] ml_parse_ie fails", __func__);
 			goto err;
diff --git a/wpa_supplicant/bss.c b/wpa_supplicant/bss.c
index 44fd0a3..053ba50 100644
--- a/wpa_supplicant/bss.c
+++ b/wpa_supplicant/bss.c
@@ -851,7 +851,7 @@ void wpa_bss_update_scan_res(struct wpa_supplicant *wpa_s,
 		if (ml_ie) {
 			struct wpa_ml_ie_parse ml;
 
-			if (ml_parse_ie(&ml_ie[3], ml_ie[1] - 1, &ml, NULL) == 0) {
+			if (ml_parse_ie(&ml_ie[3], ml_ie[1] - 1, &ml, NULL, NULL) == 0) {
 				os_memcpy(bss->aa, ml.ml_addr, ETH_ALEN);
 				wpa_printf(MSG_DEBUG,
 					"ML: BSS[" MACSTR "] set AA[" MACSTR "]",
diff --git a/wpa_supplicant/sme.c b/wpa_supplicant/sme.c
index d684127..51d92cb 100644
--- a/wpa_supplicant/sme.c
+++ b/wpa_supplicant/sme.c
@@ -227,7 +227,7 @@ static struct wpabuf * sme_auth_build_sae_commit(struct wpa_supplicant *wpa_s,
 			wpa_printf(MSG_DEBUG, "ML: clearing STA ML IE");
 			bss->dot11MultiLinkActivated = 0;
 		} else {
-			if (ml_parse_ie(ml_ie, ml_ie_len, &ml, NULL) != 0) {
+			if (ml_parse_ie(ml_ie, ml_ie_len, &ml, NULL, NULL) != 0) {
 				bss->dot11MultiLinkActivated = 0;
 			} else {
 				bss->dot11MultiLinkActivated = 1;
