Index: linux-5.4.211/include/uapi/linux/wireguard.h
===================================================================
--- linux-5.4.211.orig/include/uapi/linux/wireguard.h
+++ linux-5.4.211/include/uapi/linux/wireguard.h
@@ -180,6 +180,7 @@ enum wgpeer_attribute {
 	WGPEER_A_TX_BYTES,
 	WGPEER_A_ALLOWEDIPS,
 	WGPEER_A_PROTOCOL_VERSION,
+	WGPEER_A_PUNCH,
 	__WGPEER_A_LAST
 };
 #define WGPEER_A_MAX (__WGPEER_A_LAST - 1)
