Index: linux-5.4.281/drivers/video/fbdev/core/fbcon.c
===================================================================
--- linux-5.4.281.orig/drivers/video/fbdev/core/fbcon.c
+++ linux-5.4.281/drivers/video/fbdev/core/fbcon.c
@@ -367,6 +367,7 @@ static int get_color(struct vc_data *vc,
 
 static void fb_flashcursor(struct work_struct *work)
 {
+#if 0
 	struct fb_info *info = container_of(work, struct fb_info, queue);
 	struct fbcon_ops *ops = info->fbcon_par;
 	struct vc_data *vc = NULL;
@@ -397,6 +398,8 @@ static void fb_flashcursor(struct work_s
 	ops->cursor(vc, info, mode, get_color(vc, info, c, 1),
 		    get_color(vc, info, c, 0));
 	console_unlock();
+#endif
+	return;
 }
 
 static void cursor_timer_handler(struct timer_list *t)
@@ -1325,6 +1328,7 @@ static void fbcon_clear_margins(struct v
 
 static void fbcon_cursor(struct vc_data *vc, int mode)
 {
+#if 0
 	struct fb_info *info = registered_fb[con2fb_map[vc->vc_num]];
 	struct fbcon_ops *ops = info->fbcon_par;
  	int c = scr_readw((u16 *) vc->vc_pos);
@@ -1346,6 +1350,8 @@ static void fbcon_cursor(struct vc_data
 
 	ops->cursor(vc, info, mode, get_color(vc, info, c, 1),
 		    get_color(vc, info, c, 0));
+#endif
+	return;
 }
 
 static int scrollback_phys_max = 0;
