From 0bec534b6c635fa9e8d9e6b5400e9815c4816b33 Mon Sep 17 00:00:00 2001
From: zengyinbin <yinbin.zeng@gl-inet.com>
Date: Tue, 30 Sep 2025 14:11:42 +0800
Subject: [PATCH] fix netifd coredump

---
 interface.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/interface.c b/interface.c
index b3bb601..eed5337 100644
--- a/interface.c
+++ b/interface.c
@@ -769,12 +769,13 @@ interface_proto_event_cb(struct interface_proto_state *state, enum interface_pro
 
 		netifd_log_message(L_NOTICE, "Interface '%s' is now down\n", iface->name);
 		mark_interface_down(iface);
+		interface_write_resolv_conf(iface->jail);
 		if (iface->main_dev.dev)
 			device_release(&iface->main_dev);
 		if (iface->l3_dev.dev)
 			device_remove_user(&iface->l3_dev);
 		interface_handle_config_change(iface);
-		break;
+		return;
 	case IFPEV_LINK_LOST:
 		if (iface->state != IFS_UP)
 			return;
-- 
2.34.1

