--- a/bin/config_generate
+++ b/bin/config_generate
@@ -305,7 +305,7 @@
 		set system.@system[-1].hostname='OpenWrt'
 		set system.@system[-1].timezone='UTC'
 		set system.@system[-1].ttylogin='0'
-		set system.@system[-1].log_size='512'
+		set system.@system[-1].log_size='64'
 		set system.@system[-1].urandom_seed='0'
 
 		delete system.ntp
--- a/etc/init.d/boot
+++ b/etc/init.d/boot
@@ -41,8 +41,6 @@
 	grep -q pstore /proc/filesystems && /bin/mount -o noatime -t pstore pstore /sys/fs/pstore
 	[ "$FAILSAFE" = "true" ] && touch /tmp/.failsafe
 
-	#先检查mtkwifi的默认配置是否正常
-	/sbin/wifi check
 	/sbin/kmodloader
 
 	[ ! -f /etc/config/wireless ] && {
@@ -55,6 +53,14 @@
 	/bin/config_generate
 	uci_apply_defaults
 	sync
+
+	if [ -e /sys/devices/platform/11200000.xhci/RG_USB20_DISCTH ]; then
+		echo "2 0 4b0101" > /sys/devices/platform/11200000.xhci/RG_USB20_DISCTH
+	fi
+
+	if [ -e /sys/devices/platform/11190000.xhci/RG_USB20_DISCTH ]; then
+		echo "2 0 4b0101" > /sys/devices/platform/11190000.xhci/RG_USB20_DISCTH
+	fi
 	
 	# temporary hack until configd exists
 	/sbin/reload_config
--- a/lib/functions/system.sh
+++ b/lib/functions/system.sh
@@ -102,6 +102,15 @@
 	[ -n "$mac_dirty" ] && macaddr_canonicalize "$mac_dirty"
 }
 
+mmc_get_mac_binary() {
+	local part_name="$1"
+	local offset="$2"
+	local part
+
+	part=$(find_mmc_part "$part_name")
+	get_mac_binary "$part" "$offset"
+}
+
 mtd_get_mac_binary() {
 	local mtdname="$1"
 	local offset="$2"
--- a/lib/functions.sh
+++ b/lib/functions.sh
@@ -313,6 +313,25 @@
 	echo "${INDEX:+$PREFIX$INDEX}"
 }
 
+find_mmc_part() {
+	local DEVNAME PARTNAME ROOTDEV
+
+	if grep -q "$1" /proc/mtd; then
+		echo "" && return 0
+	fi
+
+	if [ -n "$2" ]; then
+		ROOTDEV="$2"
+	else
+		ROOTDEV="mmcblk*"
+	fi
+
+	for DEVNAME in /sys/block/$ROOTDEV/mmcblk*p*; do
+		PARTNAME="$(grep PARTNAME ${DEVNAME}/uevent | cut -f2 -d'=')"
+		[ "$PARTNAME" = "$1" ] && echo "/dev/$(basename $DEVNAME)" && return 0
+	done
+}
+
 group_add() {
 	local name="$1"
 	local gid="$2"
