Index: hostapd-2022-07-29-b704dc72/wpa_supplicant/events.c
===================================================================
--- hostapd-2022-07-29-b704dc72.orig/wpa_supplicant/events.c
+++ hostapd-2022-07-29-b704dc72/wpa_supplicant/events.c
@@ -1269,7 +1269,8 @@ static bool wpa_scan_res_ok(struct wpa_s
 		return false;
 	}
 
-	if (match_ssid_len == 0){
+	// target BSS Hiddenmode cp ssid set only when apcli set bssid param 
+	if (ssid->bssid_set && match_ssid_len == 0 && os_memcmp(bss->bssid, ssid->bssid, ETH_ALEN) == 0) {
 		os_memcpy(bss->ssid, ssid->ssid, ssid->ssid_len);
 		bss->ssid_len = ssid->ssid_len;
 		wpa_dbg(wpa_s, MSG_DEBUG, "Bssid match, In hiddenmode,then cp ssid set");
