diff --git a/src/ap/ctrl_iface_ap.c b/src/ap/ctrl_iface_ap.c
index e4a0f21..bce50c7 100644
--- a/src/ap/ctrl_iface_ap.c
+++ b/src/ap/ctrl_iface_ap.c
@@ -1164,7 +1164,7 @@ int hostapd_ctrl_iface_bss_tm_req(struct hostapd_data *hapd,
 
 	sta = ap_get_sta(hapd, addr);
 	if (sta == NULL) {
-		for (i = 0; i < 3; i++) {
+		for (i = 0; i < hapd->iface->interfaces->count ; i++) {
 			struct hostapd_iface *iface = hapd->iface;
 			if (iface) {
 				struct hostapd_iface *h = iface->interfaces->iface[i];
