diff --git a/src/ap/wnm_ap.c b/src/ap/wnm_ap.c
index e55140e..45f27a1 100644
--- a/src/ap/wnm_ap.c
+++ b/src/ap/wnm_ap.c
@@ -384,12 +384,15 @@ static int ieee802_11_send_bss_trans_mgmt_request(struct hostapd_data *hapd,
 		return -1;
 	}
 	int nr_num = hostapd_neighbor_count(hapd);
-	int nr_size = ETH_ALEN + 4 + 1 + 1 + 1 + 5;
+	/*nr_size = bssidlen(6), bssidinformation(4), opclass(1), ch(1), phytype(1),
+	 * optional subelement id(33) + 2 */
+	int nr_size = ETH_ALEN + 4 + 1 + 1 + 1 + 33 + 2;
 	int total_nr_size = nr_num * nr_size;
-	u8 *nr_data = os_malloc(total_nr_size);
+	u8 *nr_data = os_zalloc(total_nr_size);
 	int nr_data_len = 0;
 	if(nr_data == NULL) {
 		wpa_printf (MSG_ERROR, "Failed to allocate memory");
+		return -1;
 	} else {
 	    nr_data_len = hostapd_neighbor_insert_buffer(hapd, nr_data, total_nr_size);
 	}
@@ -432,6 +435,7 @@ static int ieee802_11_send_bss_trans_mgmt_request(struct hostapd_data *hapd,
 	res = hostapd_drv_send_action(hapd, hapd->iface->freq, 0,
 				      mgmt->da, &mgmt->u.action.category, len);
 	os_free(mgmt);
+	os_free(nr_data);
 	return res;
 }
 
