diff --git a/src/ml/ml_common.c b/src/ml/ml_common.c
index a3957b6..5755bbe 100644
--- a/src/ml/ml_common.c
+++ b/src/ml/ml_common.c
@@ -498,7 +498,6 @@ int ml_parse_ie(const u8 *ie, size_t len, struct wpa_ml_ie_parse *ml, u8 *bssid,
 		 */
 		while (frag_flag && (pos + 1) < end) {
 			u8 frag_len = 0;
-			pos += 2;
 			if (*pos != 254) {
 				wpa_printf(MSG_INFO,
 					"ML: invalid sta profile FragID(254) %u", *pos);
