diff --git a/src/ml/ml_common.c b/src/ml/ml_common.c
index 774f1e4..73ddcd9 100644
--- a/src/ml/ml_common.c
+++ b/src/ml/ml_common.c
@@ -392,13 +392,15 @@ int ml_parse_ie(const u8 *ie, size_t len, struct wpa_ml_ie_parse *ml, u8 *bssid,
 					struct hostapd_data *h_hapd = h->bss[0];
 					struct sta_info *sta = NULL;
 					sta = ap_get_sta(h_hapd, profile->addr);
-					if (sta)
-						continue;
-					sta = ap_sta_add(h_hapd, profile->addr);
 					if (sta) {
-						sta->flags |= WLAN_STA_AUTH | WLAN_STA_ASSOC;
-						os_memset(hapd->profiles[ml->prof_num].addr, 0, ETH_ALEN);
-						os_memcpy(hapd->profiles[ml->prof_num].addr, profile->addr, ETH_ALEN);
+						wpa_printf(MSG_INFO, "STA "MACSTR" found\n", MAC2STR(profile->addr));
+					} else {
+						sta = ap_sta_add(h_hapd, profile->addr);
+						if (sta) {
+							sta->flags |= WLAN_STA_AUTH | WLAN_STA_ASSOC;
+							os_memset(hapd->profiles[ml->prof_num].addr, 0, ETH_ALEN);
+							os_memcpy(hapd->profiles[ml->prof_num].addr, profile->addr, ETH_ALEN);
+						}
 					}
 				}
 			}
