Index: hostapd-2022-07-29-b704dc72/src/ap/sta_info.c
===================================================================
--- hostapd-2022-07-29-b704dc72.orig/src/ap/sta_info.c
+++ hostapd-2022-07-29-b704dc72/src/ap/sta_info.c
@@ -426,6 +426,7 @@ void hostapd_free_stas(struct hostapd_da
 
 	while (sta) {
 		prev = sta;
+        hostapd_drv_sta_deauth(hapd, sta->addr, WLAN_REASON_PREV_AUTH_NOT_VALID);
 		if (sta->flags & WLAN_STA_AUTH) {
 			mlme_deauthenticate_indication(
 				hapd, sta, WLAN_REASON_UNSPECIFIED);
