diff --git a/wpa_supplicant/ctrl_iface.c b/wpa_supplicant/ctrl_iface.c
index 8679183..fd2d085 100644
--- a/wpa_supplicant/ctrl_iface.c
+++ b/wpa_supplicant/ctrl_iface.c
@@ -1463,6 +1463,13 @@ done:
 	return ret;
 }
 
+#ifdef CONFIG_TESTING_OPTIONS
+static void wpas_ctrl_iface_send_commit_frame(struct wpa_supplicant *wpa_s, char * buf )
+{
+	wpa_s->reuseAPCommitScalar = atoi(buf);
+	wpa_printf(MSG_DEBUG, "Re-use ApCommit Scalar is %d\n",wpa_s->reuseAPCommitScalar);
+}
+#endif
 
 static int wpa_supplicant_ctrl_iface_wps_check_pin(
 	struct wpa_supplicant *wpa_s, char *cmd, char *buf, size_t buflen)
@@ -12390,6 +12397,10 @@ char * wpa_supplicant_ctrl_iface_process(struct wpa_supplicant *wpa_s,
 		if (ap_ctrl_iface_bss_tm_req(wpa_s, buf + 11))
 			reply_len = -1;
 #endif /* CONFIG_AP && CONFIG_WNM_AP */
+#ifdef CONFIG_TESTING_OPTIONS
+	} else if (os_strncmp(buf, "REUSEAPCOMMITSCALAR ", 20) == 0) {
+		wpas_ctrl_iface_send_commit_frame(wpa_s, buf + 20);
+#endif
 	} else if (os_strcmp(buf, "FLUSH") == 0) {
 		wpa_supplicant_ctrl_iface_flush(wpa_s);
 	} else if (os_strncmp(buf, "RADIO_WORK ", 11) == 0) {
diff --git a/wpa_supplicant/sme.c b/wpa_supplicant/sme.c
index 267a4c1..4d2b9a3 100644
--- a/wpa_supplicant/sme.c
+++ b/wpa_supplicant/sme.c
@@ -1356,6 +1356,43 @@ static int sme_check_sae_rejected_groups(struct wpa_supplicant *wpa_s,
 	return 0;
 }
 
+#ifdef CONFIG_TESTING_OPTIONS
+static int trigger_ap_commitscalar_reuse(struct wpa_supplicant *wpa_s) {
+
+	struct wpabuf *resp, *buf;
+	u8 *pos;
+	resp = wpabuf_alloc(4 + SAE_COMMIT_MAX_LEN + 100);
+	if (resp == NULL)
+	    return -1;
+	/*Initially hard coded the default group, will be modifed later based on use*/
+	wpa_s->sme.sae.group=19;
+	wpabuf_put_le16(resp, wpa_s->sme.sae.group);
+	pos = wpabuf_put(resp, wpa_s->sme.sae.tmp->prime_len);
+	if (crypto_bignum_to_bin(wpa_s->sme.sae.peer_commit_scalar, pos,
+				  wpa_s->sme.sae.tmp->prime_len,  wpa_s->sme.sae.tmp->prime_len) < 0)
+		return -1;
+	pos = wpabuf_put(resp, 2 * wpa_s->sme.sae.tmp->prime_len);
+	if (crypto_ec_point_to_bin(wpa_s->sme.sae.tmp->ec,
+				  wpa_s->sme.sae.tmp->peer_commit_element_ecc,
+				   pos, pos + wpa_s->sme.sae.tmp->prime_len) < 0)
+		return -1;
+
+	wpa_s->sme.sae.state == SAE_COMMITTED;
+	buf = wpabuf_alloc(4 + SAE_COMMIT_MAX_LEN + wpabuf_len(resp));
+	if (!buf) {
+		wpabuf_free(resp);
+		return -1;
+	}
+
+	wpa_s->sme.seq_num++;
+	sme_external_auth_build_buf(buf, resp, wpa_s->own_addr,
+			             wpa_s->sme.ext_auth_bssid, 1, wpa_s->sme.seq_num, WLAN_STATUS_SUCCESS);
+	wpa_hexdump(MSG_DEBUG, "Buff data :\n",
+			   buf, wpabuf_len(buf));
+	wpa_drv_send_mlme(wpa_s, wpabuf_head(buf), wpabuf_len(buf), 1, 0, 0);
+	return 0;
+}
+#endif
 
 static int sme_sae_auth(struct wpa_supplicant *wpa_s, u16 auth_transaction,
 			u16 status_code, const u8 *data, size_t len,
@@ -1541,11 +1578,20 @@ static int sme_sae_auth(struct wpa_supplicant *wpa_s, u16 auth_transaction,
 
 		wpabuf_free(wpa_s->sme.sae_token);
 		wpa_s->sme.sae_token = NULL;
-		if (!external)
+		if (!external) {
 			sme_send_authentication(wpa_s, wpa_s->current_bss,
 						wpa_s->current_ssid, 0);
-		else
+		} else {
+#ifdef CONFIG_TESTING_OPTIONS
+			if(wpa_s->reuseAPCommitScalar) {
+				trigger_ap_commitscalar_reuse(wpa_s);
+			} else {
+				sme_external_auth_send_sae_confirm(wpa_s, sa);
+			}
+#else
 			sme_external_auth_send_sae_confirm(wpa_s, sa);
+#endif
+		}
 		return 0;
 	} else if (auth_transaction == 2) {
 		if (status_code != WLAN_STATUS_SUCCESS)
diff --git a/wpa_supplicant/wpa_cli.c b/wpa_supplicant/wpa_cli.c
index f65b44c..7848b08 100644
--- a/wpa_supplicant/wpa_cli.c
+++ b/wpa_supplicant/wpa_cli.c
@@ -2947,6 +2947,14 @@ static int wpa_cli_cmd_bss_tm_req(struct wpa_ctrl *ctrl, int argc,
 
 #endif /* CONFIG_WNM_AP */
 
+#ifdef CONFIG_TESTING_OPTIONS
+static int wpa_cli_cmd_reuseAPCommitScalar(struct wpa_ctrl *ctrl, int argc,
+				  char *argv[])
+{
+	return wpa_cli_cmd(ctrl, "REUSEAPCOMMITSCALAR", 1, argc, argv);
+}
+#endif
+
 
 static int wpa_cli_cmd_raw(struct wpa_ctrl *ctrl, int argc, char *argv[])
 {
@@ -3931,6 +3939,10 @@ static const struct wpa_cli_cmd wpa_cli_commands[] = {
 	{ "bss_tm_req", wpa_cli_cmd_bss_tm_req, NULL, cli_cmd_flag_none,
 	  "= send BSS Transition Management Request" },
 #endif /* CONFIG_WNM_AP */
+#ifdef CONFIG_TESTING_OPTIONS
+	{ "reuseAPCommitScalar", wpa_cli_cmd_reuseAPCommitScalar, NULL, cli_cmd_flag_none,
+	  "= send frame by reusing APCommitScalar Elements" },
+#endif
 	{ "raw", wpa_cli_cmd_raw, NULL, cli_cmd_flag_sensitive,
 	  "<params..> = Sent unprocessed command" },
 	{ "flush", wpa_cli_cmd_flush, NULL, cli_cmd_flag_none,
diff --git a/wpa_supplicant/wpa_supplicant_i.h b/wpa_supplicant/wpa_supplicant_i.h
index b8d1773..842f13a 100644
--- a/wpa_supplicant/wpa_supplicant_i.h
+++ b/wpa_supplicant/wpa_supplicant_i.h
@@ -756,6 +756,9 @@ struct wpa_supplicant {
 	unsigned int assoc_freq;
 	u8 *last_con_fail_realm;
 	size_t last_con_fail_realm_len;
+#ifdef CONFIG_TESTING_OPTIONS
+	int reuseAPCommitScalar;
+#endif
 
 	/* Selected configuration (based on Beacon/ProbeResp WPA IE) */
 	int pairwise_cipher;
