diff --git a/src/ap/drv_callbacks.c b/src/ap/drv_callbacks.c
index d110de6..31c4e2a 100644
--- a/src/ap/drv_callbacks.c
+++ b/src/ap/drv_callbacks.c
@@ -1051,7 +1051,7 @@ void hostapd_event_ch_switch(struct hostapd_data *hapd, int freq, int ht,
 	}
 
 	for (i = 0; i < hapd->iface->num_bss; i++)
-		hostapd_neighbor_set_own_report(hapd->iface->bss[i]);
+		hostapd_neighbor_set_own_report(NULL, hapd->iface->bss[i]);
 
 #ifdef CONFIG_OCV
 	if (hapd->conf->ocv &&
diff --git a/src/ap/hostapd.c b/src/ap/hostapd.c
index 68528a4..48c4993 100644
--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -2161,10 +2161,11 @@ static int hostapd_setup_interface_complete_sync(struct hostapd_iface *iface,
 						 int err)
 {
 	struct hostapd_data *hapd = iface->bss[0];
-	size_t j;
+	size_t j, i;
 	u8 *prev_addr;
 	int delay_apply_cfg = 0;
 	int res_dfs_offload = 0;
+	struct hostapd_data *h_hapd_2G = NULL, *h_hapd_5G = NULL;
 
 	if (err)
 		goto fail;
@@ -2371,8 +2372,25 @@ dfs_offload:
 	if (iface->interfaces && iface->interfaces->terminate_on_error > 0)
 		iface->interfaces->terminate_on_error--;
 
-	for (j = 0; j < iface->num_bss; j++)
-		hostapd_neighbor_set_own_report(iface->bss[j]);
+	for (i = 0; i < hapd->iface->interfaces->count; i++) {
+		struct hostapd_iface *h_iface = hapd->iface->interfaces->iface[i];
+		struct hostapd_data *h_hapd = NULL;
+		if (h_iface) {
+			h_hapd = h_iface->bss[0];
+			if (i == 0 && hapd->iface->interfaces->count > 1)
+				h_hapd_2G = h_hapd;
+			else if (i == 1 && hapd->iface->interfaces->count > 2)
+				h_hapd_5G = h_hapd;
+		}
+
+		for (j = 0; j < iface->num_bss; j++) {
+			hostapd_neighbor_set_own_report(h_hapd, iface->bss[j]);
+			if (h_hapd_2G && h_hapd_2G != h_hapd)
+				hostapd_neighbor_set_own_report(h_hapd, h_hapd_2G);
+			if (h_hapd_5G && h_hapd_5G != h_hapd)
+				hostapd_neighbor_set_own_report(h_hapd, h_hapd_5G);
+		}
+	}
 
 	return 0;
 
diff --git a/src/ap/neighbor_db.c b/src/ap/neighbor_db.c
index 49b47d5..fffddd8 100644
--- a/src/ap/neighbor_db.c
+++ b/src/ap/neighbor_db.c
@@ -251,8 +251,7 @@ static enum nr_chan_width hostapd_get_nr_chan_width(struct hostapd_data *hapd,
 }
 #endif /* NEED_AP_MLME */
 
-
-void hostapd_neighbor_set_own_report(struct hostapd_data *hapd)
+void hostapd_neighbor_set_own_report(struct hostapd_data *h_hapd, struct hostapd_data *hapd)
 {
 #ifdef NEED_AP_MLME
 	u16 capab = hostapd_own_capab_info(hapd);
@@ -261,7 +260,7 @@ void hostapd_neighbor_set_own_report(struct hostapd_data *hapd)
 	int he = hapd->iconf->ieee80211ax && !hapd->conf->disable_11ax;
 	bool eht = he && hapd->iconf->ieee80211be && !hapd->conf->disable_11be;
 	struct wpa_ssid_value ssid;
-	u8 channel, op_class;
+	u8 channel, op_class, pref;
 	u8 center_freq1_idx = 0, center_freq2_idx = 0;
 	enum nr_chan_width width;
 	u32 bssid_info;
@@ -328,9 +327,9 @@ void hostapd_neighbor_set_own_report(struct hostapd_data *hapd)
 
 	/*
 	 * Neighbor Report element size = BSSID + BSSID info + op_class + chan +
-	 * phy type + wide bandwidth channel subelement.
+	 * phy type + wide bandwidth channel subelement + Candidate preference subelement.
 	 */
-	nr = wpabuf_alloc(ETH_ALEN + 4 + 1 + 1 + 1 + 5);
+	nr = wpabuf_alloc(ETH_ALEN + 4 + 1 + 1 + 1 + 5 + 3);
 	if (!nr)
 		return;
 
@@ -339,6 +338,17 @@ void hostapd_neighbor_set_own_report(struct hostapd_data *hapd)
 	wpabuf_put_u8(nr, op_class);
 	wpabuf_put_u8(nr, channel);
 	wpabuf_put_u8(nr, ieee80211_get_phy_type(hapd->iface->freq, ht, vht));
+	if (h_hapd && os_memcmp(h_hapd->own_addr, hapd->own_addr, ETH_ALEN) != 0)
+		pref = 255;
+	else
+		pref = 0;
+	/*
+	 * WNM bss transition candidate preference subelement may be needed to allow the
+	 * STA to choose the best AP.
+	 */
+	wpabuf_put_u8(nr, WNM_NEIGHBOR_BSS_TRANSITION_CANDIDATE);
+	wpabuf_put_u8(nr, 1);
+	wpabuf_put_u8(nr, pref);
 
 	/*
 	 * Wide Bandwidth Channel subelement may be needed to allow the
@@ -351,8 +361,12 @@ void hostapd_neighbor_set_own_report(struct hostapd_data *hapd)
 	wpabuf_put_u8(nr, center_freq1_idx);
 	wpabuf_put_u8(nr, center_freq2_idx);
 
-	hostapd_neighbor_set(hapd, hapd->own_addr, &ssid, nr, hapd->iconf->lci,
-			     hapd->iconf->civic, hapd->iconf->stationary_ap, 0);
+	if (h_hapd)
+		hostapd_neighbor_set(h_hapd, hapd->own_addr, &ssid, nr, hapd->iconf->lci,
+				hapd->iconf->civic, hapd->iconf->stationary_ap, 0);
+	else
+		hostapd_neighbor_set(hapd, hapd->own_addr, &ssid, nr, hapd->iconf->lci,
+				hapd->iconf->civic, hapd->iconf->stationary_ap, 0);
 
 	wpabuf_free(nr);
 #endif /* NEED_AP_MLME */
diff --git a/src/ap/neighbor_db.h b/src/ap/neighbor_db.h
index 2c0b7bd..1f8bd6a 100644
--- a/src/ap/neighbor_db.h
+++ b/src/ap/neighbor_db.h
@@ -19,7 +19,7 @@ int hostapd_neighbor_set(struct hostapd_data *hapd, const u8 *bssid,
 			 const struct wpabuf *nr, const struct wpabuf *lci,
 			 const struct wpabuf *civic, int stationary,
 			 u8 bss_parameters);
-void hostapd_neighbor_set_own_report(struct hostapd_data *hapd);
+void hostapd_neighbor_set_own_report(struct hostapd_data *h_hapd, struct hostapd_data *hapd);
 int hostapd_neighbor_remove(struct hostapd_data *hapd, const u8 *bssid,
 			    const struct wpa_ssid_value *ssid);
 void hostapd_free_neighbor_db(struct hostapd_data *hapd);
diff --git a/src/ap/ubus.c b/src/ap/ubus.c
index b7b8343..e6a09b3 100644
--- a/src/ap/ubus.c
+++ b/src/ap/ubus.c
@@ -1029,7 +1029,7 @@ __hostapd_bss_mgmt_enable_f(struct hostapd_data *hapd, int flag)
 
 		bss->radio_measurements[0] |=
 			WLAN_RRM_CAPS_NEIGHBOR_REPORT;
-		hostapd_neighbor_set_own_report(hapd);
+		hostapd_neighbor_set_own_report(NULL, hapd);
 		return true;
 	case BSS_MGMT_EN_BEACON:
 		flags = WLAN_RRM_CAPS_BEACON_REPORT_PASSIVE |
