Index: hostapd-2022-07-29-b704dc72/src/ap/pmksa_cache_auth.c
===================================================================
--- hostapd-2022-07-29-b704dc72.orig/src/ap/pmksa_cache_auth.c
+++ hostapd-2022-07-29-b704dc72/src/ap/pmksa_cache_auth.c
@@ -115,10 +115,11 @@ void pmksa_cache_free_entry(struct rsn_p
 	/*wpa_auth = pmksa->ctx;
 	hapd = wpa_auth->cb_ctx;*/
 
-	wpa_printf(MSG_INFO, "remove PMKID in driver, sta mac (" MACSTR "), bssid (" MACSTR ")",
-		MAC2STR(entry->spa), MAC2STR(wpa_auth->addr));
+	wpa_printf(MSG_ERROR, "remove PMKID in driver, sta mac (" MACSTR "), bssid (" MACSTR "), akmp 0x%x",
+		MAC2STR(entry->spa), MAC2STR(wpa_auth->addr), entry->akmp);
 	/*pmksa->pmksa->spa is the same as entry->spa*/
 	pmkid_entry.AddRemove = 0;      /*remove entry*/
+	pmkid_entry.akmp = entry->akmp;
 	os_memcpy(pmkid_entry.sta, entry->spa, ETH_ALEN); /* sta mld mac or legacy sta link addr*/
 	if (hapd->ml_group)
 		os_memcpy(pmkid_entry.bssid, hapd->ml_group->ml_addr, ETH_ALEN);
@@ -376,9 +377,10 @@ pmksa_cache_auth_add(struct rsn_pmksa_ca
 	if (pmksa_cache_auth_add_entry(pmksa, entry) < 0)
 		return NULL;
 #ifdef HOSTAPD_PMKID_IN_DRIVER_SUPPORT
-	wpa_printf(MSG_DEBUG, "add PMKID in driver spa:"MACSTR" aa:"MACSTR" link addr:"MACSTR"",
-	MAC2STR(spa),MAC2STR(aa),MAC2STR(link_addr));
+	wpa_printf(MSG_ERROR, "add PMKID in driver spa:"MACSTR" aa:"MACSTR" link addr:"MACSTR", akmp 0x%0x",
+		MAC2STR(spa),MAC2STR(aa),MAC2STR(link_addr), akmp);
 	pmkid_entry.AddRemove = 1;	      /*Add PMKID in driver entry*/
+	pmkid_entry.akmp = akmp;
 	os_memcpy(pmkid_entry.sta, spa, ETH_ALEN);
 	os_memcpy(pmkid_entry.bssid, aa, ETH_ALEN);
 	os_memcpy(pmkid_entry.link_addr, link_addr, ETH_ALEN);
Index: hostapd-2022-07-29-b704dc72/src/drivers/driver.h
===================================================================
--- hostapd-2022-07-29-b704dc72.orig/src/drivers/driver.h
+++ hostapd-2022-07-29-b704dc72/src/drivers/driver.h
@@ -1280,6 +1280,7 @@ struct wpa_pmkid_entry {
 	u8 link_addr[ETH_ALEN];
 	u8 pmkid[PMKID_LEN];
 	u8 AddRemove;   /*1- ADD, 0- Remove*/
+	int akmp; 		/* WPA_KEY_MGMT_* */
 };
 #endif /*HOSTAPD_PMKID_IN_DRIVER_SUPPORT*/
 
