diff --git a/src/ap/wpa_auth_ft.c b/src/ap/wpa_auth_ft.c
index 882177a..1a06749 100644
--- a/src/ap/wpa_auth_ft.c
+++ b/src/ap/wpa_auth_ft.c
@@ -2282,8 +2282,13 @@ static int wpa_ft_pull_pmk_r1(struct wpa_state_machine *sm,
 	struct tlv_list req_enc[] = {
 		{ .type = FT_RRB_PMK_R0_NAME, .len = WPA_PMK_NAME_LEN,
 		  .data = pmk_r0_name },
+#ifdef CONFIG_MTK_IEEE80211BE
+		{ .type = FT_RRB_S1KH_ID, .len = ETH_ALEN,
+		  .data = sm->dot11MultiLinkActivated ? sm->sta_ml_ie->ml_addr : sm->addr },
+#else
 		{ .type = FT_RRB_S1KH_ID, .len = ETH_ALEN,
 		  .data = sm->addr },
+#endif
 		{ .type = FT_RRB_LAST_EMPTY, .len = 0, .data = NULL },
 	};
 	struct tlv_list req_auth[] = {
@@ -3624,6 +3629,7 @@ static int wpa_ft_process_auth_req(struct wpa_state_machine *sm,
 	int use_sha384;
 	size_t pmk_r1_len, kdk_len;
 	const u8 *s1kh_id;
+	struct hostapd_data *hapd_mlo = sm->wpa_auth->cb_ctx;
 
 	*resp_ies = NULL;
 	*resp_ies_len = 0;
@@ -3860,7 +3866,7 @@ pmk_r1_derived:
 
 	if (sm->dot11MultiLinkActivated) {
 		wpa_printf(MSG_DEBUG, "FT: Add ML IE in authentication response");
-		ret = wpa_write_mlie(conf, mladdr, NULL, 0, pos, end - pos, true);
+		ret = wpa_write_mlie(conf, hapd_mlo->ml_group->ml_addr, hapd_mlo, 0, pos, end - pos, true);
 		if (ret < 0)
 			goto fail;
 		pos += ret;
@@ -4929,12 +4935,18 @@ struct ft_get_sta_ctx {
 static int ft_get_sta_cb(struct wpa_state_machine *sm, void *ctx)
 {
 	struct ft_get_sta_ctx *info = ctx;
-
+	u8 tmp_addr[ETH_ALEN] = {0};
+#ifdef CONFIG_MTK_IEEE80211BE
+	if (sm->dot11MultiLinkActivated)
+		os_memcpy(tmp_addr, sm->sta_ml_ie->ml_addr, ETH_ALEN);
+	else
+#endif
+	os_memcpy(tmp_addr, sm->addr, ETH_ALEN);
 	if ((info->s1kh_id &&
-	     os_memcmp(info->s1kh_id, sm->addr, ETH_ALEN) != 0) ||
-	    os_memcmp(info->nonce, sm->ft_pending_pull_nonce,
-		      FT_RRB_NONCE_LEN) != 0 ||
-	    sm->ft_pending_cb == NULL || sm->ft_pending_req_ies == NULL)
+		os_memcmp(info->s1kh_id, tmp_addr, ETH_ALEN) != 0) ||
+		os_memcmp(info->nonce, sm->ft_pending_pull_nonce,
+			FT_RRB_NONCE_LEN) != 0 ||
+		sm->ft_pending_cb == NULL || sm->ft_pending_req_ies == NULL)
 		return 0;
 
 	info->sm = sm;
