Index: hostapd-2022-07-29-b704dc72/src/ml/ml_common.c
===================================================================
--- hostapd-2022-07-29-b704dc72.orig/src/ml/ml_common.c
+++ hostapd-2022-07-29-b704dc72/src/ml/ml_common.c
@@ -688,6 +688,11 @@ int ml_group_init(struct hostapd_data *h
 
 	/* found, join it */
 	if (ml_group) {
+		wpa_printf(MSG_INFO,"old mld addr " MACSTR ", new mld addr " MACSTR " ", MAC2STR(ml_group->ml_addr),MAC2STR(mld_addr));
+		if (os_memcmp(ml_group->ml_addr, mld_addr, ETH_ALEN) != 0) {
+			wpa_printf(MSG_INFO,"mld addr changed copy new addr set");
+			os_memcpy(ml_group->ml_addr, mld_addr, ETH_ALEN);
+		}
 		/* error check */
 		for (i = 0; i < ml_group->ml_link_num; i++) {
 			link = &ml_group->links[i];
